src/log.rs

```diff
@@ -22,6 +30,127 @@ impl Operation {
                 error,
                 latency: _,
             } => op.is_some() || error.is_some(),
+            Operation::CollectStats { per_dir_stats: _, loop_stats: _ } => {
+                true // logic punted to StatCollector
+            },
         }
     }
+
+    pub fn log_str(&mut self) -> String {
+        // This unwrap seems safe, afaict. We're not cramming any user supplied strings in here.
+        serde_json::to_string(self).unwrap()
```

Comment:
From what I've been reading recently, `expect()` is almost always preferred over `unwrap()`, as you can attach a context-specific error message. I think best practise is to actually avoid both and handle all errors, but in situations like these where an error indicates a very definite & immediate bug in the code I can't see any harm.

There's a few `unwrap` calls elsewhere that could do with the same treatment. 

```suggestion
        serde_json::to_string(self).expect("failed to serialize log message")
```